### PR TITLE
Migrate away from deprecated React constructs (part 1)

### PR DESCRIPTION
**Pull request by ecraig12345**
#### Pull request checklist

- [x] Addresses an existing issue: #4613
- [ ] Include a change request file using `$ npm run change`

#### Description of changes

There are various deprecated React constructs we need to move away from in preparation for future React versions. See [this issue comment](https://github.com/OfficeDev/office-ui-fabric-react/issues/4613#issuecomment-446413003) for an outline of the work all-up.

This PR contains an initial round of (relatively) simple conversions from `componentWillReceiveProps` to `getDerivedStateFromProps`. Converting components which extend `BaseComponent` is [currently blocked](https://github.com/OfficeDev/office-ui-fabric-react/issues/4613#issuecomment-446461869), so this PR only covers things extending React.Component.

Other changes:
- Found and removed one last usage of string refs, and disabled the last lint rule override.
- Deprecated a helper function that uses the legacy context API
- Replaced some manual function refs in the Calendar component with createRef

#### Focus areas to test

(optional)

### CODE REVIEW

**Comment by khmakoto** [comment]
The Travis CI build is complaining about out of date api for the oufr package, if you could update it and add it as part of this PR it would be great.

Other than that the PR LGTM 👍

**Comment by ecraig12345** [comment]
@khmakoto Fixed the API issue

File: packages/charting/src/components/DonutChart/Arc/Arc.tsx [hunk 1]
```diff
@@ -58,11 +53,16 @@ export class Arc extends React.Component<IArcProps, { isCalloutVisible: boolean
...
-  }
+  };
 
   private _redirectToUrl(href: string | undefined): void {
     href ? (window.location.href = href) : '';
   }
 }
+
+function _updateChart(newProps: IArcProps): void {
+  newProps.arc.innerRadius(newProps.innerRadius);
```
Comment (dzearing) [inline packages/charting/src/components/DonutChart/Arc/Arc.tsx]: I know you're just moving code from one place to another... but this is a code smell.

@RajeshGoriga @samiyaakhtar props should not be mutated, ever. If you need to process them, you should process a copy of the props and set it in state (or, in the new API `getDerivedStateFromProps`, return the state values derived from the new props.)

File: packages/charting/src/components/DonutChart/Arc/Arc.tsx [hunk 2]
```diff
@@ -58,11 +53,16 @@ export class Arc extends React.Component<IArcProps, { isCalloutVisible: boolean
...
+  newProps.arc.outerRadius(newProps.outerRadius);
```
Comment (dzearing) [inline packages/charting/src/components/DonutChart/Arc/Arc.tsx]: I had left a previous comment, but misread the code. This is just calling a function, not mutating props.

**Pull request merged**